gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/test/java/io/gravitee/am/gateway/handler/root/resources/handler/webauthn/WebAuthnLoginHandlerTest.java

```diff
@@ -0,0 +1,324 @@
+
+        // init webauthn objet
+        webAuthn = webAuthn();
+
+        webAuthnLoginHandler =
+                new WebAuthnLoginHandler(factorService, factorManager, domain, webAuthn, credentialService, userAuthenticationManager);
+
+        router.route(HttpMethod.POST, "/webauthn/login")
+                .handler(SessionHandler.create(LocalSessionStore.create(vertx)))
+                .handler(BodyHandler.create())
+                .failureHandler(new LoginFailureHandler(authenticationFlowContextService));
+    }
+
+    @Test
+    public void shouldNotLogin_v1_deviceIntegrityException() throws Exception {
+        // init domain
+        WebAuthnSettings webAuthnSettings = new WebAuthnSettings();
+        webAuthnSettings.setEnforceAuthenticatorIntegrity(true);
+        webAuthnSettings.setEnforceAuthenticatorIntegrityMaxAge(10);
+        when(domain.getWebAuthnSettings()).thenReturn(webAuthnSettings);
+
+        Credential credential = new Credential();
+        credential.setCounter(0l);
+        credential.setLastCheckedAt(new Date(Instant.now().minusSeconds(60).toEpochMilli()));
+
+        when(credentialService.findByCredentialId(any(), any(), any())).thenReturn(Flowable.just(credential));
+        when(userAuthenticationManager.connectPreAuthenticatedUser(any(), any(), any())).thenReturn(Single.just(new User()));
+
+        router.route(HttpMethod.POST, "/webauthn/login")
+                .handler(rc -> {
+                    Client client = new Client();
+                    rc.session().put(ConstantKeys.PASSWORDLESS_CHALLENGE_KEY, "KidEI7a_CXoW7dKjb8JEnL_WUb2xDPWBHtCUzTo4ouAyKNzc0SbGmC0t8mh0cKHdqo_4hYc3c2Iewp_QF746rA");
+                    rc.session().put(ConstantKeys.PASSWORDLESS_CHALLENGE_USERNAME_KEY, "username");
+                    rc.put(ConstantKeys.CLIENT_CONTEXT_KEY, client);
+                    rc.next();
+                })
+                .handler(webAuthnLoginHandler)
+                .handler(rc -> rc.end());
+
+        testRequest(
+                HttpMethod.POST,
+                "/webauthn/login",
+                req -> {
+                    req.headers().set("content-type", "application/x-www-form-urlencoded");
+                    req.setChunked(true);
+                    req.write(Buffer.buffer("assertion=%7B%22id%22%3A%22QAn8i7wToriAx38jzUrVw04Cxao_y285vX2CGwyFtO8%22%2C%22rawId%22%3A%22QAn8i7wToriAx38jzUrVw04Cxao_y285vX2CGwyFtO8%22%2C%22response%22%3A%7B%22clientDataJSON%22%3A%22eyJ0eXBlIjoid2ViYXV0aG4uZ2V0IiwiY2hhbGxlbmdlIjoiS2lkRUk3YV9DWG9XN2RLamI4SkVuTF9XVWIyeERQV0JIdENVelRvNG91QXlLTnpjMFNiR21DMHQ4bWgwY0tIZHFvXzRoWWMzYzJJZXdwX1FGNzQ2ckEiLCJvcmlnaW4iOiJodHRwOi8vbG9jYWxob3N0OjgwOTIiLCJjcm9zc09yaWdpbiI6ZmFsc2V9%22%2C%22authenticatorData%22%3A%22SZYN5YgOjGh0NBcPZHZgW4_krrmihjLHmVzzuoMdl2MBAAAAAg%22%2C%22signature%22%3A%22MEUCIBfUoc-FHIliWp_6bw2c74Bnx6d62eDce0klo2O79ksjAiEA5c0JmD5o9CGacnx3bjOausUFRobNhYlJatPqRZwObn0%22%2C%22userHandle%22%3A%22%22%7D%2C%22type%22%3A%22public-key%22%7D"));
```

Comment:
Set this to a constant field, it's duplicated five times.

Reply:
FP